### PR TITLE
kubernetes-deployment

### PR DESCRIPTION
**Pull request by pennyyang17**
**Database Integration:**

- PostgreSQL is configured as the database backend for the microservice, deployed using a Kubernetes StatefulSet for reliable storage and scaling.

**All Kubernetes configurations are structured under the `k8s/` directory:**

- `deployment.yaml`: Defines the microservice's Deployment specification.
- `ingress.yaml`: Manages external access to the application via Ingress.
- `postgres.yaml`: Configures the PostgreSQL StatefulSet along with associated resources.
- `pv.yaml`: Specifies PersistentVolumes for PostgreSQL data storage.
- `secret.yaml`: Secures sensitive configuration such as database credentials.
- `service.yaml`: Exposes the microservice and PostgreSQL database internally within the cluster.

**Health Monitoring:**

- A `/health` endpoint has been added to support Kubernetes readiness and liveness probes, ensuring robust monitoring and recovery.

**Deployment Verification:**
The application has been successfully tested in a local Kubernetes environment using:
- `make cluster` to create a local cluster.
- `kubectl apply -f k8s` to deploy the application and its supporting infrastructure.

### CODE REVIEW

**Review by ian-Liaozy — COMMENTED** [review]
Looks good to me, @pennyyang17 have you verified on your local that everything works?

File: k8s/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: promotions
+  labels:
+    app: promotions
+spec:
+  replicas: 2
```
Comment (yc6714) [inline k8s/deployment.yaml:8]: do we need to replicas for now ? we can set it to 1 for simplicity, how do you think ?

File: k8s/secret.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: postgres-creds
+data:
+  database-uri: cG9zdGdyZXNxbCtwc3ljb3BnOi8vcG9zdGdyZXM6cGdzM2NyM3RAcG9zdGdyZXM6NTQzMi9wb3N0Z3Jlcw==
```
Comment (yc6714) [inline k8s/secret.yaml:7]: this one should be exactly the same as we used in `deployment.yaml`. you use `database_uri` in `deployment.yaml`. you can either choose `_` or `-`. I'm fine with both.

File: k8s/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      app: promotions
+  template:
+    metadata:
+      labels:
+        app: promotions
+    spec:
+      restartPolicy: Always
+      containers:
+        - name: promotions
+          image: cluster-registry:32000/promotions:latest
```
Comment (yc6714) [inline k8s/deployment.yaml:25]: This image should be the same used in `make cluster` command. In `Makefile`, we use `cluster-registry:5000`.

**Comment by yc6714** [comment]
Also please update `pyproject.toml` with 
```
[tool.poetry.dependencies]
python = "^3.11"
Flask = "^3.0.3"
flask-sqlalchemy = "3.1.1"
psycopg = {extras = ["binary"], version = "^3.1.19"}
retry2 = "^0.9.5"
python-dotenv = "^1.0.1"
gunicorn = "^22.0.0"
python-dateutil = "^2.9" // add this line 
```

since deployment environment is independent with our local env. if we don't add this package the service will get errors.

File: k8s/deployment.yaml (hunk 3 above)
Comment (pennyyang17) [reply k8s/deployment.yaml:25]: Gotcha! Lemme change it to the current image.

File: k8s/deployment.yaml (hunk 1 above)
Comment (pennyyang17) [reply k8s/deployment.yaml:8]: Yes that makes sense.

File: k8s/secret.yaml (hunk 2 above)
Comment (pennyyang17) [reply k8s/secret.yaml:7]: Thanks for catching that! I just changed it to database-uri in deployment.yaml

**Review by yc6714 — APPROVED** [review]

**Review by yc6714 — APPROVED** [review]

**Review by ian-Liaozy — APPROVED** [review]

**Pull request merged**